src/emmy/structure.cljc

```diff
@@ -694,6 +694,63 @@
   [f & structures]
   (sum:r:l f structures))
 
+(defn fold-chain
+  "Returns the result of accumulating all non-structural entries in `s` using the
+  supplied fold function `f` into the optional accumulator `init` (defaults
+  to `(f)`).
+
+  `f` must be a 2-argument fn of type `(accumulator, [x chain orientations]) =>
+  accumulator` responsible for merging some value `x` into the ongoing
+  accumulation. The second argument is a 3-vector containing
+
+  - the entry in the structure
+  - a vector of its 'access chain', i.e., the path you'd pass
+    to [[clojure.core/get-in]] to access the entry
+  - a vector of orientations associated with each index in the access chain
+
+  `f` should return a new instance of the accumulator.
+
+  Additional arities allow you to supply
+
+  - `init`, the initial (empty) accumulator (defaults to `(f)`)
+  - `present`, a function that will be applied to the final, aggregated
+    result (defaults to `f`)
+
+  For example:
+
+  ```clojure
+  (fold-chain
+    (fn ([] [])
+     ([acc] acc)
+     ([acc [s chain orientations]]
+      (conj acc {:s s
+                 :chain chain
+                 :orientations orientations})))
+    (s/down (s/up 1 2) (s/up 3 4)))
+
+  [{:s 1, :chain [0 0], :orientations [::s/down ::s/up]}
+   {:s 2, :chain [0 1], :orientations [::s/down ::s/up]}
+   {:s 3, :chain [1 0], :orientations [::s/down ::s/up]}
+   {:s 4, :chain [1 1], :orientations [::s/down ::s/up]}]
+  ```"
+  ([f s] (fold-chain f (f) f s))
+  ([f init s] (fold-chain f init f s))
+  ([f init present s]
+   (letfn [(walk [acc s chain orientations]
+             (if (structure? s)
+               (let [o (orientation s)]
+                 (reduce
```

Comment:
Does this mean we have one pending `reduce` per layer of "rank" in the structure? Would this be more "transducey" if we did _not_ call reduce, but instead, _generated_ the lazy sequence of (e, chain, orientations) and let that feed into a (trans|re)duction supplied elsewhere?

Reply:
yes I bet it would! the `fold` name works here because the accumulator is passed up through the levels, but your idea would let us use normal `transduce`, which would be great.